### PR TITLE
Hexadecimal numerals

### PR DESCRIPTION
**Pull request by proux01**
We add hexadecimal numerals according to the following regexp
`0[xX][0-9a-fA-F][0-9a-fA-F_]*(\.[0-9a-fA-F_]+)?([pP][+-]?[0-9][0-9_]*)?`.

This does not yet implement hexadecimal printing of primitive floats as suggested by the discussion in #11859 but is a first step towards this.

- [x] Added / updated test-suite
- [x] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

~Depends on #12163~ (merged)

Overlays:
* https://github.com/HoTT/HoTT/pull/1307 (backward compatible, merged)
* https://github.com/coq/stdlib2/pull/21
* https://github.com/coq-community/paramcoq/pull/51
* https://github.com/MetaCoq/metacoq/pull/394

### CODE REVIEW

**Review by Zimmi48 — COMMENTED** [review]
Without more documentation, it's unclear what this PR makes possible, and in particular the changelog is quite vague.

I'm also worried about the various renamings in the standard library that do not provide any (deprecated) compatibility aliases. Note that there are a number of failures in CI that are due to this (bedrock2, HoTT, stdlib2 + some plugins).

**Comment by proux01** [comment]
> the changelog is quite vague.

Right, would something like
```
Numeral notations now parse hexadecimal constants such as ``0x2a`` or ``0xb.2ap-2``.
Parsers added for nat, positive, Z, N, Q, R, primitive integers and primitive floats.
```
work?

> I'm also worried about the various renamings in the standard library that do not provide any (deprecated) compatibility aliases. Note that there are a number of failures in CI that are due to this (bedrock2, HoTT, stdlib2 + some plugins).

Indeed, let's try to keep some backward compatibility.

**Comment by Zimmi48** [comment]
Yes, much better!

**Comment by proux01** [comment]
@Zimmi48 thanks for your advice! It reduces the number of required overlays to four : the two "standard" libraries (HoTT and stdlib2) and two plugins (assuming the stack overflow from fiat-crypto is unrelated).

**Comment by Zimmi48** [comment]
Hum, out of memory errors are common, but stack overflows less so. @JasonGross do you confirm this is unrelated?

File: doc/sphinx/language/gallina-specification-language.rst [hunk 1]
```diff
@@ -74,18 +74,27 @@ Identifiers
 
 Numerals
   Numerals are sequences of digits with an optional fractional part
-  and exponent, optionally preceded by a minus sign. :n:`@int` is an integer;
-  a numeral without fractional or exponent parts. :n:`@num` is a non-negative
+  and exponent, optionally preceded by a minus sign. Hexadecimal numerals
+  start with ''0x' or ''0X'. :n:`@int` is an integer;
```
Comment (Zimmi48) [inline doc/sphinx/language/gallina-specification-language.rst]: You probably meant to use backticks instead:

```suggestion
  start with ``0x`` or ``0X``. :n:`@int` is an integer;
```

**Comment by JasonGross** [comment]
Well, the stack overflow also seems to be happening on Coq master (https://gitlab.com/coq/coq/-/jobs/489487108 is on master, right?) so it does seem likely to be unrelated.  Note that the stack overflow happens in ocamlc / ocamlopt when compiling on extracted code.  Since we do make use of the numeral definitions for conversations between numerals and strings, it would at least be possible for this PR to cause the stack overflow in fiat-crypto.  (It's possible the solution is just "increase the stack size by a bit on the command-line; I haven't seen any of these issues on fiat-crypto's CI)

**Review by Zimmi48 — COMMENTED** [review]
A warning should be emitted when a user relies on deprecated stuff. Create a deprecation warning with `Deprecation.create_warning` (defined in `interp/deprecation.mli`) or `CWarnings.create` (defined in `lib/cWarnings.mli`).

File: doc/sphinx/language/gallina-specification-language.rst (hunk 1 above)
Comment (proux01) [reply doc/sphinx/language/gallina-specification-language.rst]: Indeed, done, thanks!

**Comment by proux01** [comment]
> A warning should be emitted when a user relies on deprecated stuff. Create a deprecation warning with `Deprecation.create_warning` (defined in `interp/deprecation.mli`) or `CWarnings.create` (defined in `lib/cWarnings.mli`).

Thanks! done

**Comment by proux01** [comment]
I took the opportunity to prove the numeral notation for Q which was still unproved (after all, that's the main point of numeral notations), that is basically
```
forall q : Q, if print q = Some n then parse n = q
(not all rationals have a finite decimal representation)
forall n : Decimal.decimal, print (parse n) = normalize n
```
Amusingly enough, this proof did not unveil any bug but coming up with the right spec (the `normalize` function) wasn't entirely trivial (c.f. commit "Decimal: specify numeral notation for Q").

This being done, I consider this PR ready and won't touch it anymore. Sorry that it became so bulky but most of the volume comes from vernacular files (proving the added numeral notations) and shouldn't be a review hurdle.

CI is green except fiat-crypto failure which is likely unrelated.

**Comment by herbelin** [comment]
As far as I'm concerned, I'm fine with the PR. Does anyone want to add something, or assign?

**Comment by herbelin** [comment]
I'm assigning. Will merge say on Friday.

**Review by herbelin — COMMENTED** [review]
This is actually a rather large PR. I had a cursory look and it looks good. Maybe would @JasonGross have a look himself, since it seems to be in large part an extension of the numeral notation mechanism?

**Comment by JasonGross** [comment]
> We add hexadecimal numerals according to the following regexp
> `0[xX][0-9a-fA-F][0-9a-fA-F_]*(.[0-9a-fA-F_]+)?([pP][+-]?[0-9][0-9_]*)?`.

Do you mean `\.` rather than `.`?

File: doc/changelog/03-notations/11948-hexadecimal.rst [hunk 2]
```diff
@@ -0,0 +1,12 @@
+- **Added:**
+  Numeral notations now parse hexadecimal constants such as ``0x2a``
+  or ``0xb.2ap-2``.  Parsers added for nat, positive, Z, N, Q, R,
```
Comment (JasonGross) [inline doc/changelog/03-notations/11948-hexadecimal.rst:3]:
```suggestion
  or ``0xb.2ap-2``.  Parsers added for :g:`nat`, :g:`positive`, :g:`Z`, :g:`N`, :g:`Q`, :g:`R`,
```
(Though actually I'm not sure about the difference between :g: vs `` cc @coq/doc-maintainers )

File: doc/changelog/03-notations/11948-hexadecimal.rst [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+  primitive integers and primitive floats.
```
Comment (JasonGross) [inline doc/changelog/03-notations/11948-hexadecimal.rst:4]: I believe the standard convention is to not have `.` between the sentence and the parenthetical
```suggestion
  primitive integers and primitive floats
```

File: doc/changelog/03-notations/11948-hexadecimal.rst [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+  (`#11948 <https://github.com/coq/coq/pull/11948>`_,
+  by Pierre Roux).
+- **Deprecated:**
+  Numeral Notation on ``Decimal.uint``, ``Decimal.int`` and
+  ``Decimal.decimal`` are replaced respectively by numeral notations
+  on ``Numeral.uint``, ``Numeral.int`` and ``Numeral.numeral``.
```
Comment (JasonGross) [inline doc/changelog/03-notations/11948-hexadecimal.rst:10]:
```suggestion
  on ``Numeral.uint``, ``Numeral.int`` and ``Numeral.numeral``
```

File: doc/sphinx/user-extensions/syntax-extensions.rst [hunk 5]
```diff
@@ -1468,26 +1468,32 @@ Numeral notations
...
+     * :n:`@ident__1 -> option Numeral.uint`
      * :n:`@ident__1 -> Z`
      * :n:`@ident__1 -> option Z`
-     * :n:`@ident__1 -> Decimal.decimal`
-     * :n:`@ident__1 -> option Decimal.decimal`
+     * :n:`@ident__1 -> Numeral.numeral`
+     * :n:`@ident__1 -> option Numeral.numeral`
+
+     .. deprecated:: 8.12
+        Numeral notations on ``Decimal.uint``, ``Decimal.int`` and
```
Comment (JasonGross) [inline doc/sphinx/user-extensions/syntax-extensions.rst:1493]: Here we say
```rst
``Decimal.uint``
```
whereas below we say
```rst
:g:`Decimal.uint`
```
Does it matter?  Should we standardize?  cc @coq/doc-maintainers

File: test-suite/output/Int63Syntax.v [hunk 6]
```diff
@@ -5,6 +5,7 @@ Check (2 + 2)%int63.
 Open Scope int63_scope.
 Check 2.
 Check 9223372036854775807.
+Check 0x1a.
```
Comment (JasonGross) [inline test-suite/output/Int63Syntax.v:8]: Does Int63 support syntax such as `0x1aP5`?  If so, please add a test for it.  And please also add tests for the capital variants, such as `0X1A`, `0x1A`, `0X1a`, and zero (`0x0`, `0x00`) and also some failing tests (`0xG`, `00x1a`, `0x`)

File: test-suite/output/SearchPattern.out [hunk 7]
```diff
@@ -13,35 +13,39 @@ Nat.testbit: nat -> nat -> bool
 Nat.eqb: nat -> nat -> bool
 (use "About" for full details on implicit arguments)
 Nat.two: nat
-Nat.one: nat
 Nat.zero: nat
+Nat.one: nat
```
Comment (JasonGross) [inline test-suite/output/SearchPattern.out:17]: Can we fix `SearchPattern` to sort its results or something so that there's not so much churn?  (cc @herbelin ?)

File: test-suite/output/ZSyntax.v [hunk 8]
```diff
@@ -15,3 +16,8 @@ Check (Z.of_nat 0 = 0%Z).
 (* Submitted by Pierre Casteran *)
 Require Import Arith.
 Check (0 + Z.of_nat 11)%Z.
+
+(* Check hexadecimal printing *)
+Definition to_num_int n := Numeral.IntHex (Z.to_hex_int n).
+Numeral Notation Z Z.of_num_int to_num_int : Z_scope.
+Check 42%Z.
```
Comment (JasonGross) [inline test-suite/output/ZSyntax.v:23]: Let's also check negatives and 0
```suggestion
Check 42%Z.
Check (-42)%Z.
Check 0%Z.
```

File: theories/Init/Prelude.v [hunk 9]
```diff
@@ -32,13 +34,13 @@ Declare ML Module "string_notation_plugin".
 (* Parsing / printing of decimal numbers *)
 Arguments Nat.of_uint d%dec_uint_scope.
 Arguments Nat.of_int d%dec_int_scope.
-Numeral Notation Decimal.uint Decimal.uint_of_uint Decimal.uint_of_uint
+Numeral Notation Numeral.uint Numeral.uint_of_uint Numeral.uint_of_uint
   : dec_uint_scope.
```
Comment (JasonGross) [inline theories/Init/Prelude.v:38]: Do we also want `hex_uint_scope` registered by default?  What about `hex_nat_scope`?

File: theories/Numbers/DecimalFacts.v [hunk 10]
```diff
@@ -10,13 +10,26 @@
...
+Require Import Decimal Arith.
 
 Lemma uint_dec (d d' : uint) : { d = d' } + { d <> d' }.
 Proof.
  decide equality.
 Defined.
 
+Lemma int_dec (d d' : int) : { d = d' } + { d <> d' }.
+Proof.
+case d; clear d; intro d; case d'; clear d'; intro d'.
```
Comment (JasonGross) [inline theories/Numbers/DecimalFacts.v:22]: Slight preference for all proof bodies to be indented by two spaces.

File: theories/Numbers/DecimalQ.v [hunk 11]
```diff
@@ -0,0 +1,562 @@
...
+    Proofs that conversions between decimal numbers and [Q]
+    are bijections. *)
+
+Require Import Decimal DecimalFacts DecimalPos DecimalN DecimalZ QArith.
+
+Lemma of_to (q:Q) :
+  match to_decimal q with
+  | None => True
+  | Some d => of_decimal d = q
+  end.
```
Comment (JasonGross) [inline theories/Numbers/DecimalQ.v:22]: These kinds of lemmas are hard to `apply`, I'd much prefer the default correctness property to have type `forall q d, to_decimal q = Some d -> of_decimal d = q`.

File: theories/Numbers/DecimalQ.v [hunk 12]
```diff
@@ -0,0 +1,562 @@
...
+         set (m := match u with Nil => Pos zero | _ => _ end).
+         assert (H : m = Neg u); [|rewrite H; clear m H].
+         { now revert Hu''; unfold m; case u. }
+         now rewrite (proj2 (Nat.ltb_ge _ _) Hpe').
+Qed.
+
+Lemma to_of (d:decimal) :
+  to_decimal (of_decimal d) = Some (dnorme d)
+  \/ to_decimal (of_decimal d) = Some (dnormf d).
+Proof.
```
Comment (JasonGross) [inline theories/Numbers/DecimalQ.v:344]: It might be nice to know when one happens vs when the other happens.  That is, to classify `to_decimal (of_decimal d)` uniquely.  I could see this being too hard, though.

File: theories/Numbers/HexadecimalFacts.v [hunk 13]
```diff
@@ -0,0 +1,354 @@
...
+case d; clear d; intro d; case d'; clear d'; intro d'.
+- case (uint_dec d d').
+  + now intros->; left.
+  + now intro H; right; intro H'; apply H; injection H'.
+- now right.
+- now right.
+- case (uint_dec d d').
+  + now intros->; left.
+  + now intro H; right; intro H'; apply H; injection H'.
+Qed.
```
Comment (JasonGross) [inline theories/Numbers/HexadecimalFacts.v:31]: Why not just do `Scheme Equality uint. Scheme Equality int.` in the files in which these inductives are defined?

File: theories/Numbers/HexadecimalFacts.v [hunk 14]
```diff
@@ -0,0 +1,354 @@
...
+    | Df d => Df Nil
+    end
+  | S n =>
+    match u with
+    | Nil => Nil
+    | D0 d | D1 d | D2 d | D3 d | D4 d | D5 d | D6 d | D7 d | D8 d | D9 d
+    | Da d | Db d | Dc d | Dd d | De d | Df d =>
+      nth n d
+    end
+  end.
```
Comment (JasonGross) [inline theories/Numbers/HexadecimalFacts.v:112]: I feel like this is crying out for a function `uint_split_ind : forall (P : uint -> Type) (Pnil : P Nil) (PD : forall (D : uint -> uint) (u : uint), P u -> P (D u)) u, P u` which splits off the head from the rest of the constructor.  Alternatively, what about parameterizing the inductive over the finite type which enumerates its digits, and then doing all of this construction parametrically?  But perhaps this is all beyond the scope of this PR.

File: theories/Numbers/HexadecimalN.v [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+(** * HexadecimalN
+
+    Proofs that conversions between hexadecimal numbers and [N]
+    are bijections *)
+
+Require Import Hexadecimal HexadecimalFacts HexadecimalPos PArith NArith.
+
+Module Unsigned.
+
+Lemma of_to (n:N) : N.of_hex_uint (N.to_hex_uint n) = n.
```
Comment (JasonGross) [inline theories/Numbers/HexadecimalN.v:20]: Slight preference for module contents to all be indented by two spaces.

File: theories/Numbers/HexadecimalString.v [hunk 16]
```diff
@@ -0,0 +1,286 @@
...
+(*         *     GNU Lesser General Public License Version 2.1          *)
+(*         *     (see LICENSE file for the text of the license)         *)
+(************************************************************************)
+
+Require Import Hexadecimal Ascii String.
+
+(** * Conversion between hexadecimal numbers and Coq strings *)
+
+(** Pretty straightforward, which is precisely the point of the
+    [Decimal.int] datatype. The only catch is [Decimal.Nil] : we could
```
Comment (JasonGross) [inline theories/Numbers/HexadecimalString.v:16]: Do you mean
```suggestion
    [Hexadecimal.int] datatype. The only catch is [Hexadecimal.Nil] : we could
```
?

**Review by JasonGross — CHANGES_REQUESTED** [review]
I have at least skimmed through all of this PR.  As Hugo said, there is a lot of code.  I'm a fan of this change, and I like how it looks overall.  I think the changes I suggest for the test-suite and the changelog should be implemented before this is merged; feel free to take or leave the suggestions I've made about the code itself.

File: doc/changelog/03-notations/11948-hexadecimal.rst (hunk 2 above)
Comment (proux01) [reply doc/changelog/03-notations/11948-hexadecimal.rst]: done

File: doc/changelog/03-notations/11948-hexadecimal.rst (hunk 3 above)
Comment (proux01) [reply doc/changelog/03-notations/11948-hexadecimal.rst]: Makes sense

File: doc/sphinx/user-extensions/syntax-extensions.rst (hunk 5 above)
Comment (proux01) [reply doc/sphinx/user-extensions/syntax-extensions.rst]: Let's be uniform anyway.

File: test-suite/output/Int63Syntax.v (hunk 6 above)
Comment (proux01) [reply test-suite/output/Int63Syntax.v]: > Does Int63 support syntax such as 0x1aP5? 

No, integer numeral notations don't support exponents, even non negative. This could be extended but the change being non negligible, this probably belongs to a separate PR. Good point though!

> also add tests for the capital variants

right, done

File: theories/Init/Prelude.v (hunk 9 above)
Comment (proux01) [reply theories/Init/Prelude.v:38]: Maybe, to be perfectly honest, I'm not 100% sure I understand what this notations are for?

File: theories/Numbers/DecimalFacts.v (hunk 10 above)
Comment (proux01) [reply theories/Numbers/DecimalFacts.v:22]: Indeed, indentation is no longer uniform in this file. Original choice in this file was one space. I agree it's strange but let's respect it.

File: theories/Numbers/DecimalQ.v (hunk 12 above)
Comment (proux01) [reply theories/Numbers/DecimalQ.v:344]: This is heuristic (go for the most compact representation). The heuristic is likely complete but proving it would indeed require some work.
Moreover, I'm currently implementing the numeral notation extension discussed in #12035 which could enable further simplifications here, so I'm definitely not willing to invest too much here.

File: theories/Numbers/DecimalQ.v (hunk 11 above)
Comment (proux01) [reply theories/Numbers/DecimalQ.v]: Right, thanks!

File: theories/Numbers/HexadecimalFacts.v (hunk 13 above)
Comment (proux01) [reply theories/Numbers/HexadecimalFacts.v]: Done, thanks!

File: theories/Numbers/HexadecimalFacts.v (hunk 14 above)
Comment (proux01) [reply theories/Numbers/HexadecimalFacts.v:98]: I guess the right solution would be to map everything to lists and use already existing results on them rather than redefining everything here. I'll try to offer a later cleanup PR along this idea.

File: theories/Numbers/HexadecimalN.v (hunk 15 above)
Comment (proux01) [reply theories/Numbers/HexadecimalN.v:20]: Since this is derived from `DecimalN.v` which had no indentation, I keep it as it is.

File: theories/Numbers/HexadecimalString.v (hunk 16 above)
Comment (proux01) [reply theories/Numbers/HexadecimalString.v]: Indeed, thanks!

**Comment by proux01** [comment]
> Do you mean `\.` rather than `.`?

Indeed, fixed.

Thanks for the review!

File: theories/Init/Prelude.v (hunk 9 above)
Comment (JasonGross) [reply theories/Init/Prelude.v:38]: I thought some people might want to write decimal constants.  These are not very important, but I think we might actually want a `hex_nat_scope` for printing, no?

File: theories/Numbers/DecimalFacts.v (hunk 10 above)
Comment (JasonGross) [reply theories/Numbers/DecimalFacts.v]: Hm, I think one space is very rare in the standard library, I'd rather reindent the whole file in this PR, but will not push super-strongly for it here.

File: theories/Init/Prelude.v (hunk 9 above)
Comment (proux01) [reply theories/Init/Prelude.v:38]: So, I added a `hex_nat_scope`, unfortunatly none of the combinations of `Open Scope` or `Close Scope` I tried could activate the hexadecimal printer (parsing seem correctly affected but there seem to be no impact on printing). Could this be a bug?

File: theories/Numbers/DecimalFacts.v (hunk 10 above)
Comment (proux01) [reply theories/Numbers/DecimalFacts.v]: Ok, let's make this PR even larger.

File: theories/Init/Prelude.v (hunk 9 above)
Comment (JasonGross) [reply theories/Init/Prelude.v:38]: Indeed, reported as #12159

**Comment by herbelin** [comment]
#12163 was merged. This needs a rebase.

File: theories/Init/Prelude.v (hunk 9 above)
Comment (proux01) [reply theories/Init/Prelude.v:47]: This is (somewhat) fixed. A test case has been added at the end of `test-suite/output/NatSyntax.v`.

**Comment by proux01** [comment]
> #12163 was merged. This needs a rebase.

Thansk, rebased, CI green.

All comments from @JasonGross are now addressed (thanks again!).

File: test-suite/output/SearchPattern.out (hunk 7 above)
Comment (herbelin) [reply test-suite/output/SearchPattern.out:27]: Do you have a sorting order in mind?

Comment (JasonGross) [reply test-suite/output/SearchPattern.out:27]: Alphanumeric?  Really, any sorting order that is stable under insertion is fine (that is, inserting things should not reorder the things already in the map), and ideally a fast sorting method, if sorting is a bottleneck.

**Comment by herbelin** [comment]
Ready to merge, right?

File: test-suite/output/RealSyntax.out [hunk 17]
```diff
@@ -20,3 +20,18 @@ eq_refl : -1e-4 = -1e-4
      : -1e-4 = -1e-4
 eq_refl : -0.50 = -0.50
      : -0.50 = -0.50
+eq_refl : -26 = -26
+     : -26 = -26
+eq_refl : 2860 / IZR (BinIntDef.Z.pow_pos 2 8) = 2860 / IZR (Z.pow_pos 2 8)
```
Comment (JasonGross) [inline test-suite/output/RealSyntax.out:25]: Seems a little bit strange to me that printing displays this weird format; is this intended?

**Review by JasonGross — APPROVED** [review]
I have not re-read this carefully, but I did read most of it before, and I am looking forward to getting it in, and I think it's ready for merge.  (If the printing is not as intended, that can be fixed in a follow-up PR.)

File: test-suite/output/RealSyntax.out (hunk 17 above)
Comment (proux01) [reply test-suite/output/RealSyntax.out:25]: Indeed, the hexadecimal printer for R is not implemented. Maybe not worth investing too much into that OCaml code if it gets replaced by a Coq version in #12218

**Comment by herbelin** [comment]
Then, I'll merge tomorrow.

File: test-suite/output/SearchPattern.out (hunk 7 above)
Comment (herbelin) [reply test-suite/output/SearchPattern.out:27]: Actually, there is already an ordering, based on a heuristics computing size and number of distinct symbols. It would be easy to tweak it by modifyng line 161 of `vernac/search.ml`  to something like:
```
 let compare (gref1,_,_,p1) (gref2,_,_,p2) =
    let n = pervasives_compare p1 p2 in
    if n = 0 then some_relevant_order gref1 gref2 else n
```
But I guess, you would need first to turn the full references into qualified identifiers before sorting them, and there is order ready on qualified identifiers. So, maybe for a next time...

**Comment by herbelin** [comment]
@ejgallego, @zimmi48: when is it best to merge this PR knowing that there are overlays and a risk of CI unstably for some time?

**Comment by ejgallego** [comment]
> @ejgallego, @Zimmi48: when is it best to merge this PR knowing that there are overlays and a risk of CI unstably for some time?

It would have been good to milestone the PR a bit earlier so we could have discussed it; @Zimmi48 what do you think ? I guess we can still consider it for 8.12 as we are giving a couple of PRs that have not changed in the last weeks a bit more time.

For merging, I'd say, merge as soon as possible!

**Comment by ejgallego** [comment]
> For merging, I'd say, merge as soon as possible!

Sorry my message was not clear, I mean, merge as soon as possible if @Zimmi48 is OK with having the PR in 8.12 .

**Comment by herbelin** [comment]
Too late...

**Comment by ejgallego** [comment]
> Too late...

No worries, sorry my message was not clear; not a big deal if there is some problem we can always revert.

**Comment by Zimmi48** [comment]
I'm OK with the merge strategy.

**Comment by ejgallego** [comment]
> I'm OK with the merge strategy.

Ok, thanks @Zimmi48 , and sorry for the trouble I caused. Then let's proceed to request the overlays to be merged.

**Comment by jfehrle** [comment]
The following productions in basic.rst were not handled correctly.  Anything after `insertprodn` is automatically replaced by our new tool `doc_grammar`.  Since these are lexical descriptions that don't appear in the .mlg files, the productions should be added in `common.edit_mlg`, the order adjusted in `orderedGrammar` and then update the range of non-terminals for `insertprodn`.

We haven't rolled out this tool for general use yet.  I'll fix this in master as part of the ltac chapter update.  @Zimmi48 would you keep an eye out for this in doc changes?

```
  .. insertprodn numeral digit

  .. prodn::
     digit ::= 0 .. 9
     digitu ::= {| @digit | _ }
     hexdigit ::= {| 0 .. 9 | a..f | A..F }
     hexdigitu ::= {| @hexdigit | _ }
     decnum ::= @digit {* @digitu }
     hexnum ::= {| 0x | 0X } @hexdigit {* @hexdigitu }
     num ::= {| @decnum | @hexnum }
     sign ::= {| + | - }
     int ::= {? - } @num
     decimal ::= @decnum {? . {+ @digitu } } {? {| e | E } {? @sign } @decnum }
     hexadecimal ::= @hexnum {? . {+ @hexdigitu } } {? {| p | P } {? @sign } @decnum }
     numeral ::= {? - } {| @decimal | @hexadecimal }
```

**Comment by jfehrle** [comment]
Also this doesn't seem right.  -1 and -2 should be semantically rejected while 0x1 and 0x2 should be accepted.  What is an example that accepts `0x1`?

```
2 subgoals

  ============================
  True

subgoal 2 is:
 False
x < all: swap -1 -2.
2 subgoals

  ============================
  False

subgoal 2 is:
 True

x < all: swap 0x1 0x2.
Toplevel input, characters 14-15:
> all: swap 0x1 0x2.
>               ^
Error:
Syntax error: [tactic:ltac_use_default] expected after [tactic:tactic] (in [vernac:tactic_command]).
```

The grammar is:

```
| "swap" int_or_var int_or_var

int_or_var: [
| integer

integer: [
| bigint
]

bigint: [
| NUMERAL
| test_minus_nat "-" NUMERAL
]
```

and I expected NUMERAL would correspond to `num ::= {| @decnum | @hexnum }`.

Looks like the negative sign is not applied in the `test_minus_nat "-" NUMERAL` production in this routine:

```
let my_to_nat_string ?loc ispos s =
  match NumTok.Unsigned.to_nat s with
  | Some n -> n
```

**Comment by jfehrle** [comment]
Is was intentional to use decimal in the `Pnn' suffix for a hex number?

```
     hexadecimal ::= @hexnum {? . {+ @hexdigitu } } {? {| p | P } {? @sign } @decnum }
```

**Comment by JasonGross** [comment]
> -1 and -2 should be semantically rejected 

I believe negative numbers count from the tail off the goals list, as in Python list indexing.

> It was intentional to use decimal in the `Pnn' suffix for a hex number?

Yes, I believe so.  It seems reasonable to me to always encode exponents in decimal.

**Comment by jfehrle** [comment]
@JasonGross The conversion is broken.  Look at the code.  

> It was intentional to use decimal in the `Pnn' suffix for a hex number?

Let @proux01 speak for himself.

**Comment by proux01** [comment]
> The following productions in `basic.rst` were not handled correctly. Anything in the `prodn` after `insertprodn` is automatically replaced by our new tool `doc_grammar`. Since these are lexical descriptions that don't appear in the .mlg files, the productions should be added in `common.edit_mlg`, the order adjusted in `orderedGrammar` and then update the range of non-terminals for `insertprodn`.

Sorry, I didn't know about that.

> We haven't rolled out this tool for general use yet. I'll fix this in master as part of the ltac chapter update.

Thanks, feel free to ask for any help.

> Also this doesn't seem right. -1 and -2 should be semantically rejected

The doc of `swap` tells "If either num or num is negative then goals are counted from the end of the focused goal list" so this might be a bug in the doc which would be `swap int int` rather than `swap num num`.

> 0x1 and 0x2 should be accepted

This indeed looks strange, I should have a closer look.

> I expected NUMERAL would correspond to num ::= {| @decnum | @hexnum }

This should be `{| @decimal | @hexadecimal }` (i.e. unsigned but not necessarily integer).

> It was intentional to use decimal in the `Pnn' suffix for a hex number?

Yes, as noted by Jason, this corresponds to the format pervasive among many languages. The binary exponent after `p` is given in decimal not hexadecimal.

**Comment by Zimmi48** [comment]
> @Zimmi48 would you keep an eye out for this in doc changes?

I can try.

>this might be a bug in the doc which would be swap int int rather than swap num num.

This is a more general bug of the doc. In most places where `num` is used, it represents a possibly negative integer.
